### PR TITLE
[feat][fn] Add numListenerThreads configuration parameter in WorkerConfig

### LINKED ISSUE #8575: When a task is stopped or deleted, it will cause all other tasks to stop consuming data
**Describe the bug**
Runtime ： Thread Runtime
broker number ：3
bookie number ：6 
functionWorker number ： 6
task number: 423

**When I call the following api, all other tasks will stop consuming data.**
https://pulsar.apache.org/admin/v3/sinks/{tenant}/{namespace}/{sinkName}/stop
**When all tasks stop consuming data, we can see the following stack information.**
image
image
**And the topic stats:**
image

The reason for this problem is that there is only one PulsarClient in a functionWorker instance, and the number of externalExecutorProvider thread pools in PulsarClient is 1, so all consumers share the same listener executor. Therefore, when one or more tasks stoped, the listener executor will be blocked, and finally other tasks will stop consuming data.

I think there should be at least two solutions:
1. Add "numListenerThreads" configuration in WorkerConfig.
2. Remove the message listener in PulsarSource, and add a thread in PulsarSource to execute the consumer's receive method to take message. 

There is a problem with scheme 1: when you stop or delete a task, you will see the following in the stack
image

How do you like it? Or are there any other, more graceful, solutions to this problem?

### PR DESCRIPTION
**Pull request by xche**
Fixes #8575

Motivation

The reason for this problem is that there is only one PulsarClient in a functionWorker instance, and the number of externalExecutorProvider thread pools in PulsarClient is 1, so all consumers share the same listener executor. Therefore, when one or more tasks stoped, the listener executor will be blocked, and finally other tasks will stop consuming data.

### Modifications

1. The same subscriber uses the same thread pool
2. Adding “numListenerThreads” configuration parameter in WorkerConfig
3. When PulsarSource is closed then clear queue to prevent listener executor from being blocked

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

- [x] `doc-not-needed`

### CODE REVIEW

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/thread/ThreadRuntimeFactory.java
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/thread/ThreadRuntimeFactory.java:82]: Why are we setting the number of listener threads here to be one?  I think this config should also be respected with using the other runtimes as well.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PushPulsarSource.java
Comment (jerrypeng) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PushPulsarSource.java:82]: This code change doesn't seem relevant to this PR

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java
Comment (jerrypeng) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:105]: This code change doesn't seem relevant to this PR

**Review by jerrypeng — CHANGES_REQUESTED** [review]
There seems to be some changes that doesn't seem relevant to this PR

**Review by freeznet — DISMISSED** [review]

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/thread/ThreadRuntimeFactory.java
Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/thread/ThreadRuntimeFactory.java]: This is not resolved

**Comment by xche** [comment]
> @xche you added a config at the worker level but you PR only applies it to the ThreadRuntime and NOT the rest of the runtimes available.

**Comment by sijie** [comment]
@jerrypeng Can you review this again?

**Review by jerrypeng — DISMISSED** [review]

**Comment by codelipenghui** [comment]
@xche Could you please help check the failed tests?

**Comment by xche** [comment]
> @xche Could you please help check the failed tests?

@codelipenghui  The  failed tests related to this submission has been processed.  Could you please review this PR again?

**Review by sijie — APPROVED** [review]

**Comment by sijie** [comment]
@xche Can you rebase to the latest master?

**Comment by xche** [comment]
> @xche Can you rebase to the latest master?

@sijie rebase done

**Comment by codelipenghui** [comment]
@xche Sorry for the late response, we have fixed a couple of flaky tests in the master branch, could you please try to rebased to the master branch and resolve the conflicts? It's better to have this one in 2.10.

**Comment by tisonkun** [comment]
Rebased. @codelipenghui you may give another round.

Since we have 3 approvals already (dismissed now though), I'm going to merge this patch if CI passed and no more objection.

**Review by tisonkun — APPROVED** [review]

**Comment by tisonkun** [comment]
It seems that since the codebase evolves a lot, function-related tests failed on this patch.

Close as stale. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**
